python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,447 @@
+            "array5": InputType.array(pa.int64()),
+        },
+    ]
+
+    function_output_types = [
+        pa.int64(),
+        pa.int64(),
+        pa.int64(),
+        pa.int64(),
+    ]
+
+    function_docs = [
+        unary_doc,
+        binary_doc,
+        ternary_doc,
+        varargs_doc
+    ]
+
+    functions = [
+        unary_function,
+        binary_function,
+        ternary_function,
+        varargs_function
+    ]
+
+    function_inputs = [
+        [
+            pa.array([10, 20], pa.int64())
+        ],
+        [
+            pa.array([10, 20], pa.int64()),
+            pa.array([2, 4], pa.int64())
+        ],
+        [
+            pa.array([10, 20], pa.int64()),
+            pa.array([2, 4], pa.int64()),
+            pa.array([5, 10], pa.int64())
+        ],
+        [
+            pa.array([2, 3], pa.int64()),
+            pa.array([10, 20], pa.int64()),
+            pa.array([3, 7], pa.int64()),
+            pa.array([20, 30], pa.int64()),
+            pa.array([5, 10], pa.int64())
+        ]
+    ]
+
+    for name, \
+        in_types, \
+        out_type, \
+        doc, \
+        function, \
+        input in zip(function_names,
+                     function_input_types,
+                     function_output_types,
+                     function_docs,
+                     functions,
+                     function_inputs):
+        expected_output = function(*input)
+        register_scalar_function(
+            name, doc, in_types, out_type, function)
+
+        func = pc.get_function(name)
+        assert func.name == name
+
+        result = pc.call_function(name, input)
+        assert result == expected_output
+
+
+def test_udf_input():
+    def unary_scalar_function(scalar):
+        return pc.call_function("add", [scalar, 1])
+
+    # validate function name
+    doc = {
+        "summary": "test udf input",
+        "description": "parameters are validated"
+    }
+    in_types = {"scalar": InputType.scalar(pa.int64())}
+    out_type = pa.int64()
+    with pytest.raises(TypeError):
+        register_scalar_function(None, doc, in_types,
+                                 out_type, unary_scalar_function)
+
+    # validate function
+    with pytest.raises(TypeError, match="Object must be a callable"):
+        register_scalar_function("none_function", doc, in_types,
+                                 out_type, None)
+
+    # validate output type
+    expected_expr = "out_type must be a DataType, not None"
+    with pytest.raises(TypeError, match=expected_expr):
+        register_scalar_function("output_function", doc, in_types,
+                                 None, unary_scalar_function)
+
+    # validate input type
+    expected_expr = r'in_types must be a dictionary of InputType'
+    with pytest.raises(TypeError, match=expected_expr):
+        register_scalar_function("input_function", doc, None,
+                                 out_type, unary_scalar_function)
```

Comment:
What happens if `in_types` is a dictionary of something else?

Reply:
Idea was to let the Python throw an error, or are you proposing for a meaningful error by checking the passed values?

Reply:
I guess you're suggesting a test case for that too? Is it?

Reply:
Yes, I was suggesting a test case for that, just to make sure we don't crash instead of raising an exception.

Reply:
Sure, I will add one.